query/iterator.gen.go.tmpl

```diff
@@ -410,11 +411,75 @@ func (itr *{{$k.name}}SortedMergeIterator) pop() (*{{$k.Name}}Point, error) {
 type {{$k.name}}SortedMergeHeap struct {
 	opt   IteratorOptions
 	items []*{{$k.name}}SortedMergeHeapItem
+	// if each input comes from a unique single time series, we can make a shortcut.
+	// detection of the shortcut introduces some overhead but it gets significant
+	// performance improvement in cases like SELECT * FROM m GROUP BY *
+	fast bool
+}
+
+func (h *{{$k.name}}SortedMergeHeap) detectFast() {
+       for _, item := range h.items {
+               if item.itr.Stats().SeriesN != 1 {
+                       return
+               }
+       }
+
+       hasDup := false
+       s := make([]*{{$k.name}}SortedMergeHeapItem, len(h.items))
+       copy(s, h.items)
+
+       less := func(i, j int) bool {
+               x, y := s[i].point, s[j].point
+               ret := strings.Compare(x.Name, x.Name)
+               if ret == +1 {
+                       return true
+               } else if ret == -1 {
+                       return false
+               }
+               ret = strings.Compare(x.Tags.ID(), y.Tags.ID())
+               if ret == +1 {
+                       return true
+               } else if ret == -1 {
+                       return false
+               }
+               hasDup = true
+               return false
+       }
+       if h.opt.Ascending {
+               less = func(i, j int) bool {
+                       x, y := s[i].point, s[j].point
+                       ret := strings.Compare(x.Name, x.Name)
+                       if ret == -1 {
+                               return true
+                       } else if ret == +1 {
+                               return false
+                       }
+                       ret = strings.Compare(x.Tags.ID(), y.Tags.ID())
+                       if ret == -1 {
+                               return true
+                       } else if ret == +1 {
+                               return false
+                       }
+                       hasDup = true
+                       return false
+               }
+       }
```

Comment:
```suggestion
        less := func(i, j int) bool {
                x, y := s[i].point, s[j].point
                ret := strings.Compare(x.Name, x.Name)
                if ret != 0 {
                        // TT
                        // ret | == -1 | h.opt.Ascending | result
                        //  1  | false |  false          | true
                        // -1  | true  |  false          | false
                        //  1  | false |  true           | false
                        // -1  | true  |  true           | true
                        return ret == -1 == h.opt.Ascending
                }

                ret = strings.Compare(x.Tags.ID(), y.Tags.ID())
                if ret != 0 {
                        // TT
                        // ret | == -1 | h.opt.Ascending | result
                        //  1  | false |  false          | true
                        // -1  | true  |  false          | false
                        //  1  | false |  true           | false
                        // -1  | true  |  true           | true
                        return ret == -1 == h.opt.Ascending
                }
                hasDup = true
                return false
        }

```

Reply:
@foobar what do you think about this change? It will DRY the code up a bit. I'm not sure if I have the condition correct, so you should probably check it works 😄

Reply:
neat!  Then how about merging the two returns, like this?:

```go
ess := func(i, j int) bool {
                x, y := s[i].point, s[j].point
                ret := strings.Compare(x.Name, x.Name)
                if ret == 0 {        
                        ret = strings.Compare(x.Tags.ID(), y.Tags.ID())
                }
                if ret != 0 {
                        // TT
                        // ret | == -1 | h.opt.Ascending | result
                        //  1  | false |  false          | true
                        // -1  | true  |  false          | false
                        //  1  | false |  true           | false
                        // -1  | true  |  true           | true
                        return ret == -1 == h.opt.Ascending
                }
                hasDup = true
                return false
        }
```